File: day5/part2.c
#include <stdio.h>
#include <stdlib.h>

#define MAX_RANGES 100000

typedef struct {
    int start;
    int end;
} Range;

int compare(const void *a, const void *b) {
    return ((Range *)a)->start - ((Range *)b)->start;
}

int main(void) {
    FILE *file = fopen("input.txt", "r");
    if (!file) {
        perror("input.txt");
        return 1;
    }

    Range ranges[MAX_RANGES];
    int count = 0;

    while (1) {
        char line[128];
        if (!fgets(line, sizeof(line), file)) break;
        if (line[0] == '\n') break;

        sscanf(line, "%d-%d", &ranges[count].start, &ranges[count].end);
        count++;
    }

    fclose(file);

    if (count == 0) {
        printf("0\n");
        return 0;
    }

    qsort(ranges, count, sizeof(Range), compare);

    int total = 0;
    int currentStart = ranges[0].start;
    int currentEnd = ranges[0].end;

    for (int i = 1; i < count; i++) {
        if (ranges[i].start > currentEnd + 1) {
            total += currentEnd - currentStart + 1;
            currentStart = ranges[i].start;
            currentEnd = ranges[i].end;
        } else if (ranges[i].end > currentEnd) {
            currentEnd = ranges[i].end;
        }
    }

    total += currentEnd - currentStart + 1;
    printf("%d\n", total);

    return 0;
}

File: day5/part2.py
from __future__ import annotations
from dataclasses import dataclass
from typing import List


@dataclass(frozen=True)
class Range:
    start: int
    end: int


def parse_ranges(path: str) -> List[Range]:
    ranges: List[Range] = []

    with open(path, "r", encoding="utf-8") as f:
        for line in f:
            line = line.strip()
            if not line:
                break
            start, end = map(int, line.split("-"))
            ranges.append(Range(start, end))

    return ranges


def count_union_size(ranges: List[Range]) -> int:
    if not ranges:
        return 0

    sorted_ranges = sorted(ranges, key=lambda r: r.start)

    total = 0
    current_start = sorted_ranges[0].start
    current_end = sorted_ranges[0].end

    for r in sorted_ranges[1:]:
        if r.start > current_end + 1:
            total += current_end - current_start + 1
            current_start, current_end = r.start, r.end
        else:
            current_end = max(current_end, r.end)

    total += current_end - current_start + 1
    return total


def main() -> None:
    ranges = parse_ranges("input.txt")
    print(count_union_size(ranges))


if __name__ == "__main__":
    main()
